modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java

```diff
@@ -1414,6 +1414,7 @@ else if (this.leaderId.isEmpty()) {
     }
 
     private void becomeLeader() {
+        this.fsmCaller.onBeforeLeaderStart();
```

Comment:
This is not the only place where node could become a leader.
See `org.apache.ignite.raft.jraft.core.NodeImpl#handleTransferTimeout`

Reply:
If we want to call it here, we must call it after 
```
Requires.requireTrue(this.state == State.STATE_CANDIDATE, "Illegal state: " + this.state);
```

Reply:
> This is not the only place where node could become a leader.
See org.apache.ignite.raft.jraft.core.NodeImpl#handleTransferTimeout
Fixed.

> If we want to call it here, we must call it after
Requires.requireTrue(this.state == State.STATE_CANDIDATE, "Illegal state: " + this.state);
Nope, onBeforeLeaderStart is called when node is about to become a leader, however it's not guaranteed that it will happen, thus no preconditions expected.